core/src/nucleus/ribosome/callback/links_utils.rs

```diff
@@ -0,0 +1,82 @@
+use context::Context;
+use futures::executor::block_on;
+use holochain_core_types::{
+    entry::{entry_type::EntryType, Entry},
+    error::HolochainError,
+    link::Link,
+};
+use holochain_wasm_utils::api_serialization::validation::LinkDirection;
+use nucleus::actions::get_entry::get_entry;
+use std::sync::Arc;
+
+pub fn get_link_entries(
+    link: &Link,
+    context: &Arc<Context>,
+) -> Result<(Entry, Entry), HolochainError> {
+    let base_address = link.base();
+    let target_address = link.target();
+    let base = block_on(get_entry(&context, base_address.clone()))?.ok_or(
+        HolochainError::ErrorGeneric(String::from("Base for link not found")),
+    )?;
+    let target = block_on(get_entry(&context, target_address.clone()))?.ok_or(
+        HolochainError::ErrorGeneric(String::from("Target for link not found")),
+    )?;
+    Ok((base, target))
+}
+
+pub struct LinkDefinitionPath {
+    pub zome_name: String,
+    pub entry_type_name: String,
+    pub direction: LinkDirection,
+    pub tag: String,
+}
+
+pub fn find_link_definition_in_dna(
+    base_type: &EntryType,
+    tag: &String,
+    target_type: &EntryType,
+    context: &Arc<Context>,
+) -> Option<LinkDefinitionPath> {
+    let dna = context.get_dna().expect("No DNA found?!");
+    match base_type {
+        EntryType::App(app_entry_type) => dna
+            .get_entry_type_def(&app_entry_type)
+            .expect("Found entry type that is not defined in DNA?!")
```

Comment:
I don't think these can be expect statements, because doesn' that mean that someone could craft a validation request that would cause your machine to panic?  I think this function has to be able to return a HolochainError that can be passed back to it's caller (`validate_link_entry()`) no?

Reply:
I started writing why I thought this is not the case but I now think you are very right.
Will change :)